examples/step-67/step-67.cc

```diff
@@ -2243,6 +2561,10 @@ namespace Euler_DG
 
     euler_operator.project(ExactSolution<dim>(time), solution);
 
+#ifdef USE_ECL
+    rk_register_2 = solution;
+#endif
+
```

Comment:
Can you explain why this is needed? In other words, why do we assume that we have another copy of `solution`? Furthermore, I assume what you actually want here is `rk_register_2.copy_locally_owned_data_from(solution);`

Reply:
Maybe I can explain: For cell-centric loops, there is only one source rk_register in each RK sweep and two destination vectors, namely  the other rk_register and the solution vector. Both are updated in the `loop_cell_centric` function due to the merging of the inverse mass matrix operation with the RK integral evaluation. If we read and write in the same vector, we overwrite potentially needed data for face integrals/ evaluation of neighbor data. So, we need to have the projected initial data as source data available and the solution vector as destination vector to collect intermediate RK results. 

For face-centric loops, the solution vector is used as source vector during the RK sweep and updated separately during the element-wise inverse mass-matrix operation.

Alternatively, we could, use the rk_register directly in the `project` function, instead of the solution vector. I'm not sure which of both strategies is more intuitive.

Reply:
It is clear that the result of an RK stage cannot be written to the vector where we read from, which is the solution in the first stage. My point was that this copy is unexpected mathematically, where the assumption is that all that matters is the `solution` vector containing the data from the old time step. So I would much prefer if this could be reflected in the code to avoid surprises; I guess the question is whether to achieve that for some smart swapping of entries or an extra step in the first RK stage.

Reply:
Ok, I see your point. In the current implementation, one rk register holds the solution vector between time steps since we copy the data of the final solution vector in the last RK stage.

https://github.com/peterrum/dealii/blob/6c0333bdd584ca4b922a126f01cb29a0cbb06261/examples/step-67/step-67.cc#L1757-L1759

We cannot fully preserve the concept of three vectors, where the `solution` vector always holds the solution data: at some point, we need to use the `solution` as source and write in another destination vector. But we can switch the solution copy into the first RK stage instead of the last RK stage so that the `solution` vector contains the relevant data between different time steps. It results , however, in another read operation of the whole solution vector because we need to finish the first stage before copying the data. The current implementation avoid this additional read since the solution vector is already loaded in the last RK stage.

Reply:
@DavidSCN Let's discuss this tomorrow in person.

Reply:
@DavidSCN Thanks for the discussion today! I have updated the code accordingly.

@kronbichler What do you think about this version?